src/Serilog/Formatting/Json/JsonValueFormatter.cs

```diff
@@ -0,0 +1,325 @@
+        /// Format <paramref name="value"/> as JSON to <paramref name="output"/>.
+        /// </summary>
+        /// <param name="value">The value to format</param>
+        /// <param name="output">The output</param>
+        public void Format(LogEventPropertyValue value, TextWriter output)
+        {
+            // Parameter order of ITextFormatter is the reverse of the visitor one.
+            // In this class, public methods and methods with Format*() names use the
+            // (x, output) parameter naming convention.
+            Visit(output, value);
+        }
+
+        /// <summary>
+        /// Visit a <see cref="ScalarValue"/> value.
+        /// </summary>
+        /// <param name="state">Operation state.</param>
+        /// <param name="scalar">The value to visit.</param>
+        /// <returns>The result of visiting <paramref name="scalar"/>.</returns>
+        protected override bool VisitScalarValue(TextWriter state, ScalarValue scalar)
+        {
+            if (scalar == null) throw new ArgumentNullException(nameof(scalar));
+            FormatLiteralValue(scalar.Value, state);
+            return false;
+        }
+
+        /// <summary>
+        /// Visit a <see cref="SequenceValue"/> value.
+        /// </summary>
+        /// <param name="state">Operation state.</param>
+        /// <param name="sequence">The value to visit.</param>
+        /// <returns>The result of visiting <paramref name="sequence"/>.</returns>
+        protected override bool VisitSequenceValue(TextWriter state, SequenceValue sequence)
+        {
+            if (sequence == null) throw new ArgumentNullException(nameof(sequence));
+            state.Write('[');
+            var delim = "";
+            for (var i = 0; i < sequence.Elements.Count; i++)
+            {
+                state.Write(delim);
+                delim = ",";
+                Visit(state, sequence.Elements[i]);
+            }
+            state.Write(']');
+            return false;
+        }
+
+        /// <summary>
+        /// Visit a <see cref="StructureValue"/> value.
+        /// </summary>
+        /// <param name="state">Operation state.</param>
+        /// <param name="structure">The value to visit.</param>
+        /// <returns>The result of visiting <paramref name="structure"/>.</returns>
+        protected override bool VisitStructureValue(TextWriter state, StructureValue structure)
+        {
+            state.Write('{');
+
+            var delim = "";
+
+            for (var i = 0; i < structure.Properties.Count; i++)
+            {
+                state.Write(delim);
+                delim = ",";
+                var prop = structure.Properties[i];
+                WriteQuotedJsonString(prop.Name, state);
+                state.Write(':');
+                Visit(state, prop.Value);
+            }
+
+            if (structure.TypeTag != null)
+            {
+                state.Write(delim);
+                state.Write("_typeTag");
```

Comment:
This is the default legacy format, but now some years on and after staring at a lot of JSON.NET-encoded data, `$type` seems more natural. Thoughts?